mcs/tools/mkbundle/mkbundle.cs

```diff
@@ -1445,4 +1405,1167 @@ static string LocateFile(string default_path)
+
+			if (!File.Exists (clangPath))
+				clangPath = "";
+		}
+
+		return new ToolchainProgram ("clang.exe", clangPath, vcSDK);
+	}
+
+	static ToolchainProgram GetClangForVS15 (InstalledSDKInfo vcSDK)
+	{
+		string vcInstallDir = (vcSDK != null) ? vcSDK.InstallationFolder : "";
+		string clangPath = "";
+		if (vcInstallDir.Length != 0) {
+			string clangVersionFilePath = Path.Combine (vcInstallDir, "Auxiliary", "Build", "Microsoft.ClangC2Version.default.txt");
+			string clangVersion = File.ReadAllLines (clangVersionFilePath).ElementAt (0).Trim ();
+			string clangVersionPath = Path.Combine (vcInstallDir, "Tools", "ClangC2", clangVersion);
+
+			if (Target64BitApplication ())
+				clangPath = Path.Combine (clangVersionPath, "bin", "HostX64", "clang.exe");
+			else
+				clangPath = Path.Combine (clangVersionPath, "bin", "HostX86", "clang.exe");
+		}
+
+		return new ToolchainProgram ("clang.exe", clangPath, vcSDK);
+	}
+
+	static ToolchainProgram GetVCToolForVS12 (string tool, InstalledSDKInfo vcSDK)
+	{
+		string vcInstallDir = (vcSDK != null) ? vcSDK.InstallationFolder : "";
+		string toolPath = "";
+		if (vcInstallDir.Length != 0) {
+			if (Target64BitApplication ())
+				toolPath = Path.Combine (new string[] { vcInstallDir, "bin", "amd64", tool });
+			else
+				toolPath = Path.Combine (new string[] { vcInstallDir, "bin", tool });
+
+			if (!File.Exists (toolPath))
+				toolPath = "";
+		}
+
+		return new ToolchainProgram (tool, toolPath, vcSDK);
+	}
+
+	static ToolchainProgram GetVCToolForVS14 (string tool, InstalledSDKInfo vcSDK)
+	{
+		return GetVCToolForVS12 (tool, vcSDK);
+	}
+
+	static ToolchainProgram GetVCToolForVS15 (string tool, InstalledSDKInfo vcSDK)
+	{
+		string vcInstallDir = (vcSDK != null) ? vcSDK.InstallationFolder : "";
+		string toolPath = "";
+		if (vcInstallDir.Length != 0)
+		{
+			string toolsVersionFilePath = Path.Combine (vcInstallDir, "Auxiliary", "Build", "Microsoft.VCToolsVersion.default.txt");
+			string toolsVersion = File.ReadAllLines (toolsVersionFilePath).ElementAt (0).Trim ();
+			string toolsVersionPath = Path.Combine (vcInstallDir, "Tools", "MSVC", toolsVersion);
+
+			if (Target64BitApplication ())
+				toolPath = Path.Combine (toolsVersionPath, "bin", "HostX64", "x64", tool);
+			else
+				toolPath = Path.Combine (toolsVersionPath, "bin", "HostX86", "x86", tool);
+		}
+
+		return new ToolchainProgram (tool, toolPath, vcSDK);
+	}
+
+	static ToolchainProgram g_vcCompiler12 = null;
+	static ToolchainProgram g_vcCompiler14 = null;
+	static ToolchainProgram g_vcCompiler15 = null;
+	static ToolchainProgram GetVCCompiler ()
+	{
+		ToolchainProgram compiler = null;
+
+		const string compilerProgram = "cl.exe";
+		var vcSDK = GetInstalledVisualStudioVCSDK ();
+		if (vcSDK != null && vcSDK.ParentSDK != null) {
+			if (IsVisualStudio12 (vcSDK.ParentSDK.Version)) {
+				if (g_vcCompiler12 == null)
+					g_vcCompiler12 = GetVCToolForVS12 (compilerProgram, vcSDK);
```

Comment:
I wonder if we could initializate the global variable with this right away, i.e.

```c#
static ToolchainProgram g_vcCompiler12 = GetVCToolForVS12 ("cl.exe", vcSDK);
```

Then we could simplify these to:

```c#
if (IsVisualStudio12 (vcSDK.ParentSDK.Version))
   return g_vcCompiler12;
else if (IsVisualStudio14 (vcSDK.ParentSDK.Version))
  return g_vcCompiler14;
...
```

If the GetVCToolForVS12 is expensive we could do it with `Lazy<T>` so it's only accessed when it's needed.